### ISSUE TITLE
Any known issues with many controllers writing to the same DB?

### ISSUE THREAD

**Issue by danxmoran** [argoproj/argo#2054]
# Summary

My team is planning to run multiple Argo controllers (one per namespace) within a single cluster. We'd like to make use of the status-offload and archive features. We'd also like to run a single Argo server instance in the cluster so we can see the status of everything running in a single UI. We were thinking that we'd set up a single database for all of these systems to interact with. Are there any known issues with that sort of concurrent DB access? Is there a chance that the controllers might step on each other / pick up each other's workflows?

# Motivation

Each of the Argo controllers is going to be configured with a `kubeconfig` pointing to an auto-scaled cluster in a different GCP billing project. We expect the number of controllers will grow over time as we take on new projects. We're hoping that we can run just one Argo server instance to monitor everything.

**Comment by alexec** [comment]
Good question, and requires a detailed answer. 

## Offloading Node Status

If configured on, we will offload the large workflows (ones over 1MB). This is only after we attempted to compress it. In many common uses cases, even if you enable offloading, it will not need to be used.

The specific field we compress is `/status/nodes`. This is written into a table named `argo_workflows`. This table is primary keyed on (cluster name, namespace, name) - but you MUST configure the cluster name if you connect multiple cluster to a single database:

https://github.com/argoproj/argo/blob/master/docs/workflow-controller-configmap.yaml#L139

Unfortunately can't it is not possible to determine the cluster name automatically.

Offloaded data is delete when in no longer needed, which is 5m after the resource is deleted.

All runtime data is stored in Kubernetes. This is unchanged.

## Workflow Archive

If configured on, a workflow is written to the archive when it completes. It'll be in both the archive and Kubernetes until it is deleted from Kubernetes (e.g. `argo delete my-wf`), when it will remain in the archive.

Records in the archive are stored in `argo_archived_workflows` and primary keyed on (cluster name, uid). The cluster name is the same as above, the UID is the Kubernetes UID - this is used instead of name, namespace because name may be duplicated in the archive (it can't be in the runtime data in Kubernetes).

Again, if you want to use the same database for multiple clusters you must configure the cluster name (see above).  

Data can be delete from the archive using `argo archive delete UID`.

## Conclusion

To use a single database for multiple clusters  you MUST configure the cluster name.
